domain/unitstate/state/state.go

```diff
@@ -28,28 +28,142 @@ func NewState(factory database.TxnRunnerFactory) *State {
-		return err
+// GetUnitState returns the full unit state. The state may be
+// empty.
+// If no unit with the namw exists, a [errors.UnitNotFound] error is returned.
+func (st *State) GetUnitState(ctx context.Context, name coreunit.Name) (unitstate.RetrievedUnitState, error) {
+	db, err := st.DB()
+	if err != nil {
+		return unitstate.RetrievedUnitState{}, errors.Errorf("getting db: %w", err)
+	}
+
+	var state unitState
+	q := "SELECT &unitState.* FROM unit_state WHERE unit_uuid = $unitUUID.uuid"
+	unitStateStmt, err := st.Prepare(q, state, unitUUID{})
+	if err != nil {
+		return unitstate.RetrievedUnitState{}, errors.Errorf("preparing select unit state statement: %w", err)
+	}
+
+	var charmKVs []unitCharmStateKeyVal
+	q = `
+SELECT &unitCharmStateKeyVal.*
+FROM unit_state_charm
+WHERE unit_uuid = $unitUUID.uuid`
+	charmStateStmt, err := st.Prepare(q, unitCharmStateKeyVal{}, unitUUID{})
+	if err != nil {
+		return unitstate.RetrievedUnitState{}, errors.Errorf("preparing select unit charm state statement: %w", err)
+	}
+
+	var relationKVs []unitRelationStateKeyVal
+	q = `
+SELECT &unitRelationStateKeyVal.*
+FROM unit_state_relation
+WHERE unit_uuid = $unitUUID.uuid`
+	relationStateStmt, err := st.Prepare(q, unitRelationStateKeyVal{}, unitUUID{})
+	if err != nil {
+		return unitstate.RetrievedUnitState{}, errors.Errorf("preparing select unit relation state statement: %w", err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		id, err := st.getUnitUUIDForName(ctx, tx, name)
+		if err != nil {
+			return errors.Errorf("getting unit UUID for %q: %w", name, err)
+		}
+
+		err = tx.Query(ctx, unitStateStmt, id).Get(&state)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting unit state: %w", err)
+		}
+
+		err = tx.Query(ctx, charmStateStmt, id).GetAll(&charmKVs)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting unit charm state: %w", err)
+		}
+
+		err = tx.Query(ctx, relationStateStmt, id).GetAll(&relationKVs)
+		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+			return errors.Errorf("getting unit relation state: %w", err)
+		}
+
+		return nil
 	})
+	if err != nil {
+		return unitstate.RetrievedUnitState{}, err
+	}
+
+	unitState := unitstate.RetrievedUnitState{
+		UniterState:  state.UniterState,
+		StorageState: state.StorageState,
+		SecretState:  state.SecretState,
+	}
+	if len(charmKVs) > 0 {
+		unitState.CharmState = makeMapFromCharmUnitStateKeyVals(charmKVs)
+	}
+	if len(relationKVs) > 0 {
+		unitState.RelationState = makeMapFromRelationUnitStateKeyVals(relationKVs)
+	}
+
+	return unitState, nil
+}
+
+func (st *State) SetUnitState(ctx context.Context, as unitstate.UnitState) error {
+	if as.Name.Validate() != nil {
+		return errors.Errorf("invalid unit name: %q", as.Name)
+	}
 
+	db, err := st.DB()
 	if err != nil {
-		return "", errors.Errorf("getting unit UUID for %q: %w", name, err)
+		return errors.Errorf("getting db: %w", err)
```

Comment:
```suggestion
		return errors.Capture(err)
```
That's what we do elsewhere.

Reply:
Done